src/components/date.rs

```diff
@@ -45,11 +45,185 @@ impl<C: CalendarProtocol> Date<C> {
+        // 2. If options is not present, set options to undefined.
+        // 3. If duration.[[Years]] ≠ 0, or duration.[[Months]] ≠ 0, or duration.[[Weeks]] ≠ 0, then
+        if duration.date().years != 0.0
+            || duration.date().months != 0.0
+            || duration.date().weeks != 0.0
+        {
+            // a. If dateAdd is not present, then
+            // i. Set dateAdd to unused.
+            // ii. If calendar is an Object, set dateAdd to ? GetMethod(calendar, "dateAdd").
+            // b. Return ? CalendarDateAdd(calendar, plainDate, duration, options, dateAdd).
+            return self.calendar().date_add(self, duration, overflow, context);
+        }
+
+        // 4. Let overflow be ? ToTemporalOverflow(options).
+        // 5. Let norm be NormalizeTimeDuration(duration.[[Hours]], duration.[[Minutes]], duration.[[Seconds]], duration.[[Milliseconds]], duration.[[Microseconds]], duration.[[Nanoseconds]]).
+        // 6. Let days be duration.[[Days]] + BalanceTimeDuration(norm, "day").[[Days]].
+        let days = duration.days()
+            + TimeDuration::from_normalized(duration.time().to_normalized(), TemporalUnit::Day)?.0;
+
+        // 7. Let result be ? AddISODate(plainDate.[[ISOYear]], plainDate.[[ISOMonth]], plainDate.[[ISODay]], 0, 0, 0, days, overflow).
+        let result = self
+            .iso
+            .add_iso_date(&DateDuration::new(0f64, 0f64, 0f64, days)?, overflow)?;
+
+        Ok(Self::new_unchecked(result, self.calendar().clone()))
+    }
+
+    /// Returns a duration representing the difference between the dates one and two with a provided context.
+    ///
+    /// Temporal Equivalent: 3.5.6 `DifferenceDate ( calendar, one, two, options )`
+    #[inline]
+    pub(crate) fn internal_diff_date(
+        &self,
+        other: &Self,
+        largest_unit: TemporalUnit,
+        context: &mut C::Context,
+    ) -> TemporalResult<Duration> {
+        if self.iso.year == other.iso.year
+            && self.iso.month == other.iso.month
+            && self.iso.day == other.iso.day
+        {
+            return Ok(Duration::default());
+        }
+
+        if largest_unit == TemporalUnit::Day {
+            let days = self.days_until(other);
+            return Ok(Duration::from_date_duration(&DateDuration::new(
+                0f64,
+                0f64,
+                0f64,
+                f64::from(days),
+            )?));
+        }
+
+        self.calendar()
+            .date_until(self, other, largest_unit, context)
+    }
+
+    /// Equivalent: DifferenceTemporalPlainDate
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn diff_date(
+        &self,
+        op: bool,
+        other: &Self,
+        rounding_mode: Option<TemporalRoundingMode>,
+        rounding_increment: Option<f64>,
+        largest_unit: Option<TemporalUnit>,
+        smallest_unit: Option<TemporalUnit>,
+        context: &mut C::Context,
+    ) -> TemporalResult<Duration> {
```

Comment:
Same recommendation as the other PR. It would be better to do the sign calculation and application outside of this function, and the rounding type can just be passed as an argument.